sunscreen_math/src/metal_impl/ristrettovec.rs

```diff
@@ -0,0 +1,716 @@
+use curve25519_dalek::{edwards::EdwardsPoint, ristretto::RistrettoPoint, CannonicalFieldElement};
+use metal::Buffer;
+
+use core::slice;
+use std::{
+    mem::size_of,
+    ops::{Add, Mul, Sub},
+};
+
+use crate::metal_impl::{ScalarVec, U32Arg};
+
+use super::Runtime;
+
+pub struct RistrettoPointVec {
+    data: Buffer,
+    len: usize,
+}
+
+impl RistrettoPointVec {
+    /// Creates a new [RistrettoVec].
+    ///
+    /// # Remarks
+    /// This code assumes the following layout of curve25519-dalek datastructures:
+    /// ```rust
+    /// struct RistrettoPoint(EdwardsPoint);
+    ///
+    /// struct EdwardsPoint {
+    ///     X: FieldElement2625,
+    ///     Y: FieldElement2625,
+    ///     Z: FieldElement2625,
+    ///     T: FieldElement2625,
+    /// }
+    ///
+    /// struct FieldElement2625([u32; 10]);
+    /// ```
+    /// To achieve this layout, you must use the u32 backend.
+    ///
+    pub fn new(x: &[RistrettoPoint]) -> Self {
+        let runtime = Runtime::get();
+
+        let len = x.len();
+
+        assert_eq!(size_of::<RistrettoPoint>(), size_of::<u32>() * 40);
+        let byte_len = x.len() * size_of::<RistrettoPoint>();
+        let data = runtime.alloc(byte_len);
+
+        let mut field_vec = Self { data, len };
+
+        let data_slice = field_vec.buffer_slice_mut();
+
+        for (i, p) in x.iter().enumerate() {
+            let x = p.0.X.to_u29();
+            let y = p.0.Y.to_u29();
+            let z = p.0.Z.to_u29();
+            let t = p.0.T.to_u29();
+
+            for (j, w) in x.iter().enumerate() {
+                data_slice[(j) * len + i] = *w;
+            }
+
+            for (j, w) in y.iter().enumerate() {
+                data_slice[(j + 10) * len + i] = *w;
+            }
+
+            for (j, w) in z.iter().enumerate() {
+                data_slice[(j + 20) * len + i] = *w;
+            }
+
+            for (j, w) in t.iter().enumerate() {
+                data_slice[(j + 30) * len + i] = *w;
+            }
+        }
+
+        field_vec
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn len_bytes(&self) -> usize {
+        self.len * size_of::<RistrettoPoint>()
+    }
+
+    pub fn get(&self, index: usize) -> RistrettoPoint {
+        if index > self.len {
+            panic!("Index {index} exceeds bounds of {}", self.len);
+        }
+
+        let mut x = [0u32; 10];
+        let mut y = [0u32; 10];
+        let mut z = [0u32; 10];
+        let mut t = [0u32; 10];
+
+        let buffer_slice = self.buffer_slice();
+
+        for i in 0..10 {
+            x[i] = buffer_slice[i * self.len + index];
+        }
+
+        for i in 0..10 {
+            y[i] = buffer_slice[(i + 10) * self.len + index];
+        }
+
+        for i in 0..10 {
+            z[i] = buffer_slice[(i + 20) * self.len + index];
+        }
+
+        for i in 0..10 {
+            t[i] = buffer_slice[(i + 30) * self.len + index];
+        }
```

Comment:
We seem to be doing the same indexing here as above: is there a way to abstract it out? It might make it easier to reason about the magic numbers like 10 as well.

Reply:
Agree, a few well-named variables/consts would probably help this seem less magical/arbitrary.

Reply:
10 is the length in `u32`s a Field is in the 32-bit backend. IIRC I made `x` and such's type `Deref` into `[u32; 10]`, so I can call `len()`.